array/array.go

```
@@ -0,0 +1,253 @@
+	// by checking the capacity and calling Resize if necessary.
+	Reserve(n int)
+
+	// Resize adjusts the space allocated by b to n elements. If n is greater than b.Cap(),
+	// additional memory will be allocated. If n is smaller, the allocated memory may reduced.
+	Resize(n int)
+
+	// NewArray creates a new array from the memory buffers used
+	// by the builder and resets the Builder so it can be used to build
+	// a new array.
+	NewArray() Interface
+}
+
+type Boolean = array.Boolean
+
+type BooleanBuilder struct {
+	b *array.BooleanBuilder
+}
+
+func NewBooleanBuilder(mem memory.Allocator) *BooleanBuilder {
+	return &BooleanBuilder{
+		b: array.NewBooleanBuilder(mem),
+	}
+}
+func (b *BooleanBuilder) Retain() {
+	b.b.Retain()
+}
+func (b *BooleanBuilder) Release() {
+	b.b.Release()
+}
+func (b *BooleanBuilder) Len() int {
+	return b.b.Len()
+}
+func (b *BooleanBuilder) Cap() int {
+	return b.b.Cap()
+}
+func (b *BooleanBuilder) NullN() int {
+	return b.b.NullN()
+}
+func (b *BooleanBuilder) Append(v bool) {
+	b.b.Append(v)
+}
+func (b *BooleanBuilder) AppendValues(v []bool, valid []bool) {
+	b.b.AppendValues(v, valid)
+}
+func (b *BooleanBuilder) UnsafeAppend(v bool) {
+	b.b.UnsafeAppend(v)
+}
+func (b *BooleanBuilder) AppendNull() {
+	b.b.AppendNull()
+}
+func (b *BooleanBuilder) UnsafeAppendBoolToBitmap(isValid bool) {
+	b.b.UnsafeAppendBoolToBitmap(isValid)
+}
+func (b *BooleanBuilder) Reserve(n int) {
+	b.b.Reserve(n)
+}
+func (b *BooleanBuilder) Resize(n int) {
+	b.b.Resize(n)
+}
+func (b *BooleanBuilder) NewArray() Interface {
+	return b.NewBooleanArray()
+}
+func (b *BooleanBuilder) NewBooleanArray() *Boolean {
+	return b.b.NewBooleanArray()
+}
+
+type String struct {
+	value  string
+	length int
+	data   *array.Binary
+}
+
+func (a *String) DataType() arrow.DataType {
+	return arrow.BinaryTypes.String
+}
+func (a *String) Data() *array.Data {
+	if a.data != nil {
+		return a.data.Data()
+	}
+	return nil
```

Should this panic instead of returning `nil`?

---

I'm wondering if we should maybe just remove it? It's primarily used for slicing or it would be used for direct access to arrow buffers if we wanted to perform serialization in the future. It might be worth removing this from the interface and modifying the `Slice()` function to either be part of the interface or smarter about how it slices.

---

I would be in favor of removing it.